plugins/janus_recordplay.c

```diff
@@ -2618,6 +2710,73 @@ static void *janus_recordplay_playout_thread(void *data) {
 				}
 			}
 		}
+		if(data) {
+			if(data == session->dframes) {
+				/* First packet, send now */
```

Comment:
I don't think this is correct. The first data frame should not be sent until the timestamp says it should. You can have a recording created at time 0, and the first data message saved 15s later. The mjr parsing for data does indeed take that into account, by using `c_time` as a reference to calculate the relative time of each packet:

    p->ts = when-c_time;

where the first `p->ts` may very well not be `0`.

Reply:
I guess this can be solved by getting rid of the `first` branch of the if, and just stick to the second, where the `ts_diff` calculation uses something else (a monotonic timer variable created as soon as the streaming start, perhaps) instead of `data->prev->ts` for the first frame (where `data->prev` doesn't exist), e.g.

   ts_diff = data->ts - data->prev->ts;   --->   ts_diff = data->ts - start_time;

Reply:
Actually, since `data-ts` is calculated as `when-c_time`, it means that the timestamp are all indexed to 0.
So all we need is to use 0 the first time.

Reply:
No, because the timestamp in the packet (those 8 bytes) will be a monotonic time of when that packet was written to the file; `c_time` is the monotonic time of then the mjr file was first created. As such, `when-c_time` is the time that passed between since the recording file was created and when the first packet was written (which will often be > 0).

Reply:
Most of the times, you create a recording and no data is written to it for a while (much more likely for data, audio/video are there almost right away often).

Reply:
But in that case, we would need to correlate the `c_time` of the different files - audio, video - and assume that the one that starts first represents 0 time for the data... Is that a good assumption?
From what I can see in the streaming plugin (which is the only one generating data recordings), all files are created and header files written almost simultaneously, and `c_time` will be very close indeed.

Reply:
For audio/video it's rarely needed, as they flow continuously, so it can be pretty much safely ignored. Data channel messages are occasional, instead: just think of how they're used in the EchoTest demo, for instance, where audio and video flow all the time, and data only when you type something in.

Reply:
The recordings right now are started at the same time, even if data is only written much later.
However, in terms of scalability, it makes sense that all the `c_time` fields are kept at playout time, to make sure they align, for future use of recordings that are not started simultaneously.

So what I propose is for the `c_time` to be extracted out to the playout thread (from `get_frames` ), so that it can be used.

Reply:
Actually I just checked with a recording of the echotest, and timing of the data was 100% correct! I only sent data later, and it matched exactly the moment in audio I said I had just sent the message. There is another problem, though: Janus crashed for me when I tried playing a non-data recording after I had just closed the data one. I'll have a look at the asan traces and will provide info in a comment.

Reply:
As I've seen on my side. 
I'd much rather leave is as it is - it assumes the recordings are done simultaneously. It's simpler and easy to understand the behaviour if you feed in recordings done at different time.

Reply:
Yep, since it behaves as it should, I agree there's no need to change it :+1: 
It's a first integration anyway: in case issues appear later on, they can be fixed in time.